src/raw/mod.rs

```diff
@@ -872,19 +872,17 @@ impl<T, A: Allocator + Clone> RawTable<T, A> {
     /// This does not check if the given element already exists in the table.
     #[cfg_attr(feature = "inline-more", inline)]
     #[cfg(any(feature = "raw", feature = "rustc-internal-api"))]
-    pub fn insert_no_grow(&mut self, hash: u64, value: T) -> Bucket<T> {
-        unsafe {
-            let (index, old_ctrl) = self.table.prepare_insert_slot(hash);
-            let bucket = self.table.bucket(index);
+    pub unsafe fn insert_no_grow(&mut self, hash: u64, value: T) -> Bucket<T> {
+        let (index, old_ctrl) = self.table.prepare_insert_slot(hash);
+        let bucket = self.table.bucket(index);
 
-            // If we are replacing a DELETED entry then we don't need to update
-            // the load counter.
-            self.table.growth_left -= special_is_empty(old_ctrl) as usize;
+        // If we are replacing a DELETED entry then we don't need to update
+        // the load counter.
+        self.table.growth_left -= special_is_empty(old_ctrl) as usize;
```

Comment:
Since this function is already unsafe, we could add an extra `assume_no_deleted` argument which assumes that the table is fresh and does not contain any `DELETED` entries? This would skip the check on `old_ctrl` and might improve performance in some cases.

However it might be exposing too much of the hash table internals. What do you think?

Reply:
So `assume_no_deleted` would still be a condition there, but one that you would hope is constant-propagated?

I think it's too much of a micro-optimization, especially for exposing internals. I ran some `indexmap` benchmarks that are heavy on `rebuild_hash_table`, which is a `clear` followed by a loop of `insert_no_grow`. About 50% of that function's perf profile is simply the `bucket.write`, mostly due to cache misses. Most of the remaining time is spread around finding the insert slot, and less than 4% on updating `growth_left`. I even tried hard-coding that to `growth_left -= 1`, but the difference was lost in the noise.